File: EER 2/EER 2/xacTextEntry.m
//
//  xacTextEntry.m
//  EER
//
//  Created by Xiang 'Anthony' Chen on 5/30/13.
//  Copyright (c) 2013 hotnAny. All rights reserved.
//

#import "xacTextEntry.h"

@implementation xacTextEntry

UIImageView *imgView;
UITextView *textView;
NSMutableArray *charArray;
int ptrChar = 0;

- (id) init {
    self = [super init];
    if(self) {
        [NSTimer scheduledTimerWithTimeInterval:0.1
                                         target:self
                                       selector:@selector(updateTimeOut)
                                       userInfo:nil
                                        repeats:YES];
        
        
        _keyMap = [[NSMutableDictionary alloc] init];
        [self readKeyMap];
        _gestureMap = [[NSMutableDictionary alloc] init];
        [self initGestureMap];
        
        textView = nil;
        charArray = [[NSMutableArray alloc] initWithCapacity:TEXTLINELENGTH];
        imgView = nil;
    }
    return self;
}

- (void) update :(xacSwipe*) swipe {
    switch (_state) {
        case NONE:
            _numTouch = swipe.touchPoints.count;
            _firstSwipe = [self getCopyOf:swipe];
            [self updateVisual:_firstSwipe.gesture];
            _state = INPROGRESS;
            break;
        case INPROGRESS:
            _secondSwipe = [self getCopyOf:swipe];
            [self doTextEntry];
            _secondSwipe = _firstSwipe = nil;
            _state = DONE;
            [self updateVisual:UNKNOWN];
            break;
        case DONE:
            _state = NONE;
            break;
        default:
            break;
    }
    
    [self printState];
}

- (xacSwipe*) getCopyOf :(xacSwipe*)theSwipe {
    xacSwipe *tmpSwipe = [[xacSwipe alloc] init];
    tmpSwipe.gesture = theSwipe.gesture;
    tmpSwipe.touchPoints = theSwipe.touchPoints;
    tmpSwipe.timeStamp = theSwipe.timeStamp;
    return tmpSwipe;
}

- (void) doTextEntry {
    
    if(_firstSwipe.gesture == EAST && _secondSwipe.gesture == EAST) {
        NSString *key = @" ";
        if(ptrChar <= TEXTLINELENGTH - 1) {
            charArray[ptrChar++] = key;
        } else {
            [charArray removeObjectAtIndex:0];
            [charArray addObject:key];
        }
    }
    else if(_firstSwipe.gesture == WEST && _secondSwipe.gesture == WEST) {
        if(charArray.count > 0) {
            [charArray removeLastObject];
            ptrChar--;
        }
    }
    else {
        
        NSString *strFirstSwipe = [_gestureMap objectForKey:[NSNumber numberWithInt:_firstSwipe.gesture]];
        NSString *strSecondSwipe = [_gestureMap objectForKey:[NSNumber numberWithInt:_secondSwipe.gesture]];
        NSString *strID = [NSString stringWithFormat:@"_%d_%@_%@", _numTouch, strFirstSwipe, strSecondSwipe];
        
        NSString *key = [_keyMap objectForKey: strID];
        if(key) {
            if(ptrChar <= TEXTLINELENGTH - 1) {
                charArray[ptrChar++] = key;
            } else {
                [charArray removeObjectAtIndex:0];
                [charArray addObject:key];
            }
            //        ptrChar = (ptrChar + 1) % TEXTLINELENGTH;
        }
        NSLog(@"%@: %@", strID, key);
        
    }
    [self updateTextView];
}

- (void) updateTextView {
        
    NSString *sentence = @"";
    for(NSString *str in charArray) {
        sentence = [sentence stringByAppendingString:str];
    }
    [textView setText:sentence];
}

- (long) getCurrentTimeInMS
{
    NSTimeInterval time = ([[NSDate date] timeIntervalSince1970]);
    long digits = (long)time; // this is the first 10 digits
    int decimalDigits = (int)(fmod(time, 1) * 1000); // this will get the 3 missing digits
    long timeStamp = (digits * 1000) + decimalDigits;
    return timeStamp;
}

- (void) updateTimeOut {
    if(_firstSwipe == nil) {
        _state = NONE;
    }
    else {
        long currTime = [self getCurrentTimeInMS];
        if(currTime - _firstSwipe.timeStamp > BTWNTIMEOUT) {
            
            if(_state == INPROGRESS) {
                NSLog(@"time out!");
                if(imgView != nil) {
                    [self updateVisual:UNKNOWN];
                }
            }
            
            _state = NONE;
            _firstSwipe = nil;
        }
    }

}

- (void) printState {
    NSString* strState = @"unknown";
    switch (_state) {
        case NONE:
            strState = @"none";
            break;
        case INPROGRESS:
            strState = @"waiting for next stroke... ";
            break;
        case DONE:
            strState = @"done";
            break;
        default:
            break;
    }
    
    NSLog(@"%@", strState);
}

- (void) initGestureMap {
    [_gestureMap setObject:@"center" forKey:[NSNumber numberWithInt:CENTER]];
    [_gestureMap setObject:@"north" forKey:[NSNumber numberWithInt:NORTH]];
    [_gestureMap setObject:@"northeast" forKey:[NSNumber numberWithInt:NORTHEAST]];
    [_gestureMap setObject:@"east" forKey:[NSNumber numberWithInt:EAST]];
    [_gestureMap setObject:@"southeast" forKey:[NSNumber numberWithInt:SOUTHEAST]];
    [_gestureMap setObject:@"south" forKey:[NSNumber numberWithInt:SOUTH]];
    [_gestureMap setObject:@"southwest" forKey:[NSNumber numberWithInt:SOUTHWEST]];
    [_gestureMap setObject:@"west" forKey:[NSNumber numberWithInt:WEST]];
    [_gestureMap setObject:@"northwest" forKey:[NSNumber numberWithInt:NORTHWEST]];
}

- (void) readKeyMap {
    NSString *path = [[NSBundle mainBundle] pathForResource:@"key-mapping" ofType:@"csv"];
    NSString *strFile = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];

    NSRange textRange = [strFile rangeOfString:@"\n"];
    for ( int idxNewLine = textRange.location + 1;textRange.location != NSNotFound;)
    {
        strFile = [strFile substringFromIndex:idxNewLine];
        
        int numFeat = 3;
        NSString *strID = @"";
        
        for (int i=0; i<numFeat; i++) {
            int idxComma = [strFile rangeOfString:@","].location;
            NSString *strFeat = [strFile substringToIndex:idxComma];
            strID = [strID stringByAppendingFormat:@"_%@", strFeat];
            strFile = [strFile substringFromIndex:idxComma + 1];
        }
        
        textRange = [strFile rangeOfString:@"\n"];
        idxNewLine = textRange.location + 1;
        
        NSString *strKey = @"";
        if (textRange.location == NSNotFound) {
            strKey = strFile;
        }
        else {
            strKey = [strFile substringToIndex:idxNewLine-1];
        }
        
//        NSLog(@"%@", strKey);
        [_keyMap setObject:strKey forKey:strID];
    }
}

- (void) initVisualView :(UIView*) view {
    imgView = [[UIImageView alloc] init];
    imgView.image = [UIImage imageNamed:@"keyboard.png"];
    imgView.frame = CGRectMake(0, 0, view.frame.size.width, view.frame.size.height);
    [view addSubview:imgView];
}

- (void) initTextView :(UIView*) view {
    float W = view.frame.size.width;
    float H = view.frame.size.height;
    textView = [[UITextView alloc] initWithFrame:CGRectMake(0, H * 0.15, W, H * 0.6)];
    [textView setFont:[UIFont fontWithName:@"ArialMT" size:150 / TEXTLINELENGTH]];
    textView.textAlignment = NSTextAlignmentCenter;
    [textView setBackgroundColor:[UIColor clearColor]];
    [textView setUserInteractionEnabled:NO];
    [view addSubview:textView];
}

- (void) updateVisual :(int)swipe {
    switch (swipe) {
        case CENTER:
            imgView.image = [UIImage imageNamed:@"FGH.png"];
            break;
        case NORTH:
            imgView.image = [UIImage imageNamed:@"TYU.png"];
            break;
        case NORTHEAST:
            imgView.image = [UIImage imageNamed:@"IOP.png"];
            break;
        case EAST:
            imgView.image = [UIImage imageNamed:@"JKL.png"];
            break;
        case SOUTHEAST:
            imgView.image = [UIImage imageNamed:@"VBNM.png"];
            break;
        case SOUTH:
//            imgView.image = [UIImage imageNamed:@"VB.png"];
            if(imgView != nil) {
                [self updateVisual:UNKNOWN];
            }
            _state = NONE;
            _firstSwipe = nil;
            break;
        case SOUTHWEST:
            imgView.image = [UIImage imageNamed:@"ZXC.png"];
            break;
        case WEST:
            imgView.image = [UIImage imageNamed:@"ASD.png"];
            break;
        case NORTHWEST:
            imgView.image = [UIImage imageNamed:@"QWER.png"];
            break;
        default:
            imgView.image = [UIImage imageNamed:@"keyboard.png"];
            break;
    }

}

@end

File: Chewbie/EER/src/me/xiangchen/app/eer/xacTextEntry.java
package me.xiangchen.app.eer;

import java.util.Calendar;
import java.util.Hashtable;
import java.util.Timer;
import java.util.TimerTask;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.view.Gravity;
import android.widget.ImageView;
import android.widget.LinearLayout.LayoutParams;
import android.widget.RelativeLayout;
import android.widget.TextView;

@SuppressLint("NewApi")
public class xacTextEntry extends TimerTask{

	final int FPS = 10;
	final int BTWNTIMEOUT = 2000;
	enum EntryState {
	    NONE,
	    INPROGRESS,
	    DONE
	};

	int numTouch;
	xacSwipe firstSwipe;
	xacSwipe secondSwipe;
	EntryState state;
	Hashtable<String, String> keyMap;
	Hashtable<Integer, String> gestureMap;
	ImageView imgView;
	TextView textView;
	Timer timer;
	
	int widthWatch;
	int heightWatch;
	
	public xacTextEntry() {
		firstSwipe = null;
		secondSwipe = null;
		initGestureMap();
		initKeyMap();
		state = EntryState.NONE;
		timer = new Timer();
		timer.scheduleAtFixedRate(this, 0, 1000/FPS);
	}

	public void update(xacSwipe swipe) {
		switch (state) {
        case NONE:
            numTouch = swipe.touchPoints.size();
            try {
				firstSwipe = swipe.clone();
			} catch (CloneNotSupportedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
//            [self updateVisual:firstSwipe.gesture];
            state = EntryState.INPROGRESS;
            break;
        case INPROGRESS:
            try {
				secondSwipe = swipe.clone();
			} catch (CloneNotSupportedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
            doTextEntry();
            secondSwipe = firstSwipe = null;
            state = EntryState.DONE;
//            [self updateVisual:UNKNOWN];
            break;
        case DONE:
            state = EntryState.NONE;
            break;
        default:
            break;
		}
	}
	
	public void initVisualView(RelativeLayout layout, Context context) {
		imgView = new ImageView(context);
		imgView.setImageResource(R.drawable.keyboardapplealphabet);
		imgView.layout(0, 0, widthWatch, heightWatch);
		RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(
				LayoutParams.MATCH_PARENT,
				LayoutParams.MATCH_PARENT);
		params.addRule(RelativeLayout.ABOVE, 0);
		layout.addView(imgView, params);
	}
	
	public void initTextView(RelativeLayout layout, Context context) {
		textView = new TextView(context);
		textView.setText("?");
		textView.setGravity(Gravity.CENTER);
		textView.setTextColor(Color.BLACK);
		textView.setTextSize(16);
		textView.layout(0, 0, widthWatch, (int)(heightWatch * 0.85));
		RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(
				LayoutParams.MATCH_PARENT, 
				(int) (heightWatch * 0.85));
		params.addRule(RelativeLayout.ALIGN_TOP, 0);
		layout.addView(textView, params);
	}
	
	@SuppressLint("UseValueOf")
	private void doTextEntry() {
	    String strFirstSwipe = gestureMap.get(new Integer(firstSwipe.gesture));
	    String strSecondSwipe = (String)gestureMap.get(new Integer(secondSwipe.gesture));
	    String strID = numTouch + "," + strFirstSwipe + "," + strSecondSwipe;
	    
	    String keyEntry = (String) keyMap.get(strID);
	    keyEntry = keyEntry == null ? "NA" : keyEntry;
	    
	    System.out.println(strID + ": " + keyEntry);
	    
//	    [textView setText:key];
	    textView.setText(keyEntry);
	}
	
	public void setDimension(int w, int h) {
		widthWatch = w;
		heightWatch = h;
	}
	
	@SuppressLint("UseValueOf")
	private void initGestureMap() {
		gestureMap = new Hashtable<Integer, String>();
	    gestureMap.put(new Integer(xacSwipe.CENTER), new String("center"));
	    gestureMap.put(new Integer(xacSwipe.NORTH), new String("north"));
	    gestureMap.put(new Integer(xacSwipe.NORTHEAST), new String("northeast"));
	    gestureMap.put(new Integer(xacSwipe.EAST), new String("east"));
	    gestureMap.put(new Integer(xacSwipe.SOUTHEAST), new String("southeast"));
	    gestureMap.put(new Integer(xacSwipe.SOUTH), new String("south"));
	    gestureMap.put(new Integer(xacSwipe.SOUTHWEST), new String("southwest"));
	    gestureMap.put(new Integer(xacSwipe.WEST), new String("west"));
	    gestureMap.put(new Integer(xacSwipe.NORTHWEST), new String("northwest"));
	}
	
	private void initKeyMap() {
		keyMap = new Hashtable<String, String>();
		
		keyMap.put("1,northwest,west", "Q");
		keyMap.put("1,northwest,north", "W");
		keyMap.put("1,northwest,south", "E");
		keyMap.put("1,northwest,east", "R");
		keyMap.put("1,north,west", "T");
		keyMap.put("1,north,center", "Y");
		keyMap.put("1,north,east", "U");
		keyMap.put("1,northeast,west", "I");
		keyMap.put("1,northeast,center", "O");
		keyMap.put("1,northeast,east", "P");
		keyMap.put("1,west,west", "A");
		keyMap.put("1,west,center", "S");
		keyMap.put("1,west,east", "D");
		keyMap.put("1,center,west", "F");
		keyMap.put("1,center,center", "G");
		keyMap.put("1,center,east", "H");
		keyMap.put("1,east,west", "J");
		keyMap.put("1,east,center", "K");
		keyMap.put("1,east,east", "L");
		keyMap.put("1,southwest,west", "Z");
		keyMap.put("1,southwest,center", "X");
		keyMap.put("1,southwest,east", "C");
		keyMap.put("1,south,west", "V");
		keyMap.put("1,south,east", "B");
		keyMap.put("1,southeast,west", "N");
		keyMap.put("1,southeast,east", "M");
	}

	@Override
	public void run() {
		// TODO Auto-generated method stub
		if(firstSwipe == null) {
	        state = EntryState.NONE;
	    }
	    else {
	    	Calendar lCDateTime = Calendar.getInstance();
	        long currTime = lCDateTime.getTimeInMillis();
//	        System.out.println(currTime);
	        if(currTime - firstSwipe.timeStamp > BTWNTIMEOUT) {
	            
	            if(state ==EntryState.INPROGRESS) {
	                System.out.println("time out!");
	                if(imgView != null) {
//	                    [self updateVisual:EntryState.UNKNOWN];
	                }
	            }
	            
	            state = EntryState.NONE;
	            firstSwipe = null;
	        }
	    }
	}
	
}
